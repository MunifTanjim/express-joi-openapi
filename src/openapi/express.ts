import { Application, Router } from 'express'
import { RegisteredExpressOpenAPIPlugin } from '../plugins/types'
import { HttpMethod } from '../types'
import { OpenAPISpecification } from './index'
import { Layer, Route, StackLayer, StackName } from './types'

const regexPattern = {
  param: /\(\?:\(\[\^\\\/]\+\?\)\)/,
  path: /^\^\\\/(?:(:?[\w\\.-]*(?:\\\/:?[\w\\.-]*)*)|(\(\?:\(\[\^\\\/]\+\?\)\)))\\/,
  rootPath: /^\^\\\/\?\(\?=\\\/\|\$\)$/,
}

const getPathString = (pathRegex: RegExp, keys: Layer['keys']): string => {
  const pathRegexpSource = pathRegex.source

  let regexString = pathRegexpSource

  let regexStringExecArray = regexPattern.path.exec(regexString)

  let paramIndex = 0
  while (regexPattern.param.test(regexString)) {
    const param = `:${keys[paramIndex].name}`

    regexString = regexString.replace(regexPattern.param, param)

    paramIndex++
  }

  if (regexString !== pathRegexpSource) {
    regexStringExecArray = regexPattern.path.exec(regexString)
  }

  const pathString = regexStringExecArray![1].replace(/\\\//g, '/')

  return pathString
}

const processRoute = (
  spec: OpenAPISpecification,
  route: Route,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[]
): void => {
  const routePaths = Array.isArray(route.path) ? route.path : [route.path]

  routePaths.forEach((routePath) => {
    const path = `${
      basePath ? `${basePath}${routePath === '/' ? '' : routePath}` : routePath
    }`.replace(/:([a-zA-Z0-9_]+)/g, '{$1}')

    const methods = Object.keys(route.methods).filter(
      (method): method is HttpMethod => method !== '_all'
    )

    if (!spec.paths[path]) {
      spec.setPathItem(path, {})
    }

    for (const method of methods) {
      spec.setPathItemOperation(path, method, {
        responses: {
          default: {
            description: '',
          },
        },
      })

      for (const plugin of plugins) {
        const stashedValue = plugin.stash.find(route, method)
        if (stashedValue) {
          plugin.processRoute(plugin.specification, stashedValue, {
            path,
            method,
          })
        }
      }
    }
  })
}

export const processExpressRouters = (
  spec: OpenAPISpecification,
  app: Application | Router,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[]
): void => {
  const stack: StackLayer[] = app.stack ?? (app as Application)._router?.stack

  if (!stack) {
    spec.setPathItem(basePath, {})

    return
  }

  for (const layer of stack) {
    // terminal route
    if (layer.route) {
      processRoute(spec, layer.route, basePath, plugins)

      continue
    }

    if (
      layer.name !== StackName.ROUTER &&
      layer.name !== StackName.BOUND_DISPATCH &&
      layer.name !== StackName.MOUNTED_APP
    ) {
      continue
    }

    if (regexPattern.path.test(layer.regexp.source)) {
      const parsedPath = getPathString(layer.regexp, layer.keys)

      processExpressRouters(
        spec,
        layer.handle as Router,
        basePath + '/' + parsedPath,
        plugins
      )

      continue
    }

    if (
      !layer.path &&
      layer.regexp &&
      !regexPattern.rootPath.test(layer.regexp.source)
    ) {
      const regexPath = ' RegExp(' + layer.regexp + ') '

      processExpressRouters(
        spec,
        layer.handle as Router,
        basePath + '/' + regexPath,
        plugins
      )

      continue
    }

    processExpressRouters(spec, layer.handle as Router, basePath, plugins)
  }
}
