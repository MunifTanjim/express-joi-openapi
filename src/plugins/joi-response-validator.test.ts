import Joi from '@hapi/joi'
import express, { Application, Handler } from 'express'
import request from 'supertest'
import { ExpressOpenAPI } from '../index'
import {
  GetJoiResponseValidatorMiddleware,
  JoiResponseValidator,
  ResponseValidationError,
} from './joi-response-validator'

describe('JoiResponseValidator', () => {
  let validate: GetJoiResponseValidatorMiddleware
  let app: Application

  const errorHandler = jest.fn((err, _req, res, _next) => {
    if (err instanceof ResponseValidationError) {
      res.status(500).json({
        error: {
          message: err.message,
          segment: err.segment,
          details: err.validationError.details.map(
            ({ message, path, type }) => ({ message, path, type })
          ),
        },
      })

      return
    }

    res.status(500).json({
      error: {
        message: 'Server Error',
      },
    })
  })

  beforeEach(() => {
    const expressOpenApi = new ExpressOpenAPI()

    app = express()
    app.use(express.json())

    validate = expressOpenApi.registerPlugin(JoiResponseValidator)
  })

  test('validates headers', async () => {
    app.post(
      '/ping',
      validate({
        200: {
          headers: Joi.object({
            'x-count': Joi.number().integer().required(),
          }).unknown(),
        },
        default: {},
      }),
      (req, res) => {
        const pass = Number(req.query.pass)

        if (pass) {
          res.set('x-count', '42')
        }

        res.status(200).json({
          pong: 42,
        })
      }
    )

    app.use(errorHandler)

    const responseFail = await request(app).post('/ping?pass=0')

    expect(responseFail.status).toBe(500)
    expect(responseFail.body).toMatchSnapshot()

    const responsePass = await request(app).post('/ping?pass=1')

    expect(responsePass.status).toBe(200)
  })

  test('validates body', async () => {
    app.post(
      '/ping',
      validate({
        200: {
          body: Joi.object({
            pong: Joi.number().integer().required(),
          }).unknown(),
        },
        default: {},
      }),
      (req, res) => {
        const pass = Number(req.query.pass)

        res.status(200).json({
          pong: pass ? 42 : 'forty-two',
        })
      }
    )

    app.use(errorHandler)

    const responseFail = await request(app).post('/ping?pass=0')

    expect(responseFail.status).toBe(500)
    expect(responseFail.body).toMatchSnapshot()

    const responsePass = await request(app).post('/ping?pass=1')

    expect(responsePass.status).toBe(200)
  })
})
