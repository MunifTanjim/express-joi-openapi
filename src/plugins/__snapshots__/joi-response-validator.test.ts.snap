// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoiResponseValidator validates body 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"pong\\" must be a number",
        "path": Array [
          "pong",
        ],
        "type": "number.base",
      },
    ],
    "message": "\\"pong\\" must be a number",
    "segment": "body",
  },
}
`;

exports[`JoiResponseValidator validates headers 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "message": "\\"x-count\\" is required",
        "path": Array [
          "x-count",
        ],
        "type": "any.required",
      },
    ],
    "message": "\\"x-count\\" is required",
    "segment": "headers",
  },
}
`;
